### ISSUE TITLE
Layering entities

### ISSUE THREAD

**Issue by LNow** [kieler/klayjs#7]
Is it possible to use this library to auto-place elements like this ones?
![screenshot from 2016-01-23 18 20 29](https://cloud.githubusercontent.com/assets/829210/12531552/03a012c0-c1fe-11e5-9460-250934ba8daa.png)

Additionaly, it is possible to collapse groups (entities) while maintaining links like it's done here: http://live.yworks.com/yfiles-for-html/1.3/demos/complete/demo.yfiles.graph.incrementalhierarchicgrouping/index.html?

**Comment by uruuru** [comment]
If you refer to the _linkage_ of incoming edges on the left side of a node and the corresponding outgoing edge on the right side of a node, it's not possible in the current version. However, we have the feature implemented on a branch and I hope to merge it into the master soon. 

Regarding your second question: this library performs layout solely, you would have to implement the logic to collapse/expand groups yourself. Determining a layout of the hierarchical diagram with edges crossing the hierarchy boundaries is possible though.

**Comment by LNow** [comment]
What if I define _ports_ ie. for each field one input and one output port, and force its position?

Collapsing/expanding is pretty easy. I'm worried about edges. If I collapse one node that have 3 outgoing edges to another node I should have still 3 outgoing edges from this node. However when I collapse the second node, those 3 edges should be combined to one.
Is it possible?

**Comment by uruuru** [comment]
Ah sure, you can set `portConstraints: "FIXED_POS"` for the node and specify explicit positions for each port. Note that port positions are relative to the node's top-left corner. 
In my previous answer I thought one step further where you allow the fields to re-arrange in order to reduce the number of edge crossings.

I guess you would have to replace the 3 edges by a single edge as soon as you detect that both involved nodes are collapsed. I don't believe we have a feature that does something like this automatically.

**Comment by LNow** [comment]
It doesn't work well. Locking ports in place seems doesn't work at all, but it's probably because i'm drawing nodes and ports as html object and edges as svg paths.

**Comment by uruuru** [comment]
Can you share a screenshot illustrating what you mean by 'not well'?

**Comment by LNow** [comment]
@uruuru i have found a bug in my code. Locking ports works as it should.
Now I'm trying to find a way to display ports in one line (as a one single field).

**Comment by uruuru** [comment]
I'm not sure what you mean.

**Comment by LNow** [comment]
So as you know, I want to have Input links on the left side of the node, and output ones on the right side. To achieve this I have to add 2 ports for each field (one for input and one for output).
Because in the input JSON structure I have 2 ports per field, I have duplicate fields (fields are created based on ports information).
To eliminate this issue, I have to find a way to treat 2 consecutive ports as singe field and at the same time locate those ports in the correct spots.

**Comment by le-cds** [comment]
There are currently two ways to do so.
1. Calculating the port coordinates beforehand and locking them in place. As I understand, that's pretty much what you are already doing.
2. Specifying the order of ports on each side. If the left side of a node contains as many ports as the right side, this will effectively yield the same results.

The downside of both approaches is that the layout algorithm cannot reorder the ports to reduce the number of edge crossings in the graph. We have experimented with ways to properly support this use case in the past, but I don't think we have merged them into the production code. I trust @uruuru will correct me if I'm wrong there. ;)

**Comment by LNow** [comment]
@le-cds Yes i'm calculating port coordinates upfront. Thank's to you I've came up with fairly simple algorithm for my problem.
Because have ports ordered in JSON like that:

```
- INPUT_FIELD1
- OUTPUT_FIELD1
- INPUT_FIELD2
- OUTPUT_FIELD2
...
- INPUT_FIELDn
- OUTPUT_FIELDn
```

The only thing that I had to do was hide fields caused by having output ports (to eliminate fields duplications) and change output ports theirs Y coordinate.

I don't need ( and I don't want ) ports reordering feature. I'm making a web visualization of something created in 3rd party software. If developer placed fields in certain order - I want to have exactly the same order.
